src/machine.rs

```diff
@@ -249,6 +249,37 @@ impl<'m> Machine<'m> {
         Ok(())
     }
 
+    /// Delete the machine.
+    ///
+    /// Deletes the machine, cleaning up all associated resources.
+    #[instrument(skip_all)]
+    pub async fn delete(self) -> Result<(), Error> {
+        let vm_id = self.config.vm_id();
+
+        if self.is_running() {
```

Comment:
As I have written before, the best way is not to check the resource before doing any operation on it but rather do the operation and handle the errors from the operation, if any. Otherwise, it's racy. The resource could be gone between the time you check it and when you do the operation on it.

Having said that, we should keep a state enum in `Machine` and keep it updated. Also that's the one we should check in here rather than the heavy impl of `is_running` you're adding.

Reply:
> Having said that, we should keep a state enum in `Machine` and keep it updated. Also that's the one we should check in here rather than the heavy impl of `is_running` you're adding.

I'm not certain introducing artificial states will do any good. Also it's not currently possible to add because `connect` function does not know how to calculate the actual state of running machine.

Reply:
> I'm not certain introducing artificial states will do any good

It's not artificial. Its our idea of the state (which could be wrong in problematic cases but that's why I was talking about monitoring the process).

>  Also it's not currently possible to add because connect function does not know how to calculate the actual state of running machine.

Well connect needs to find out the PID anyway so when it does, it knows that the state is shutdown. I know this doesn't cover the "not-yet-started-but-created" state we've now but the solution there is simple: we launch the process as part of the `start` method. `start` should be able to do it's down after a shutdown anyway so this change is needed in any case.

Reply:
we could wrap Machine::pid into Option<pid>

`create` will return Machine { pid: None, ... }
`start` will update the pid into Some(pid)
`connect` will require users to bring the pid with them

Reply:
Sure, I was thinking of an enum:

```rust
enum MachineState {
    Off,
    On(pid: Pid),
}
```

and expose a getter for that in the API. `Option<Pid>` will serve the same right now but we'll have to switch to the enum if/when we need another state in the future.